geps/gep-1713.md

```diff
@@ -0,0 +1,220 @@
+
+## API
+
+We propose adding a new `infrastructure` stanza to the `spec` field of a Gateway. Within `infrastructure` there is an `attachTo` field of type `LocalObjectReference`. Although the use of `LocalObjectReference` allows users to attach to any `kind`, this GEP only defines the behaviour of attaching a Gateway to another Gateway.
+
+
+See [GEP-1867](https://github.com/kubernetes-sigs/gateway-api/pull/1868) for more use cases of `infrastructure`.
+
+#### Go
+
+```go
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+
+type GatewayInfrastructure struct {
+  // AttachTo allows the Gateway to associate itself with another resource.
+  // A common scenario is to reference another Gateway which marks
+  // this Gateway a child of another.
+  AttachTo LocalObjectReference `json:"attachTo"`
+}
+```
+
+#### YAML
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+spec:
+  gatewayClassName: example
+  infrastructure:
+    attachTo:
+      name: parent-gateway
+      kind: Gateway
+      group: gateway.networking.k8s.io
+  listeners:
+  - name: metrics
+    port: 8080
+    protocol: HTTP
+```
+
+### Semantics
+
+#### Gateway Attaching
+
+Gateways that _do not_ reference another Gateway (using `spec.infrastructure.attachTo`) are considered "primary" Gateways. A "child" Gateway is one where `spec.infrastructure.attachTo` field is set to a "primary" Gateway - making the "primary" Gateway a "parent". A "sibling" is a Gateway that shares a "parent" with another "child" Gateway. A "child" Gateway can only attach to Gateways in the same namespace.
+
+Failure to attach a "child" Gateway to another MUST result in the "child" Gateway setting their `Accepted` condition to `False`.
+
+This MUST happen in the following scenarios:
+- Attaching to a non-existing Gateway
+
+If a Gateway's `Accepted` condition is `False` then the conditions 'Programmed' and 'Ready' (if supported) MUST be set to `False`.
+
+Invalid Gateways MUST be rejected by the implementation during creation or update. This MUST happen in the following scenarios:
+- Attaching to another "child" Gateway
+- Attaching to oneself (self-referential)
+- Attaching to a Gateway with a different `GatewayClassName`
+
+#### Route Attaching
+
+Routes MUST be able to specify a "child" Gateways as a `parentRef` and make use of the fields in `ParentReference` to help target a specific listener.
+
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: httproute-example
+spec:
+  parentRefs:
+  - name: child-gateway
+    sectionName: metrics
+```
+
+#### Merging Spec Fields
+
+##### GatewayClassName
+
+The `GatewayClassName` MUST be set on "child" and "parent" Gateways and be equivalent.
```

Comment:
If I understand correctly, the Gateway will assume the properties of the specified GatewayClass at a given moment in time.  If the GatewayClass changes between the parent Gateway creation and the child gateway creation, this would lead to a mismatch between the parameters indicated by the parent and child.

Reply:
> If the GatewayClass changes between the parent Gateway creation and the child gateway creation, this would lead to a mismatch between the parameters indicated by the parent and child.

I don't think it's valid to change the gatewayClass after a Gateway has been created.

Reply:
I think it would be handy to drop the requirement on the "child" Gateway but it would make the validation a bit harder. Right now this field is required so the API server will reject it if it's not set. 

Changing it would require a webhook or dealing with it when reconciling (thus losing making the feedback loop async)